Enable PagedAttention through Pallas

wonjoolee95 opened the pull request:
Enable PagedAttention through Pallas

Test plan:
```
root@1fdc3324aeef:/pytorch/xla# python test/test_pallas.py PallasTest.test_paged_attention_wrapper
.
----------------------------------------------------------------------
Ran 1 test in 2.209s

OK
root@1fdc3324aeef:/pytorch/xla# python test/test_pallas.py PallasTest.test_paged_attention_wrapper_with_dynamo
.
----------------------------------------------------------------------
Ran 1 test in 2.114s

OK
```

Todo as follow-ups:
- Add unit test for Dynamo
- Enable all other parameters for jax.experimental.pallas.ops.tpu.paged_attention.paged_attention_kernel

JackCaoG commented on torch_xla/experimental/custom_kernel.py:
```diff
@@ -366,3 +385,29 @@ def flash_attention_non_xla(q: torch.Tensor,
+                        v_pages: torch.Tensor, lengths: torch.Tensor,
+                        page_indices: torch.Tensor,
+                        pages_per_compute_block: int):
+  return flash_attention(q, k, v, causal=causal)
```
is this a typo? It still call flash_attnetion

wonjoolee95 replied:
Yep a typo haha. Updated.

miladm commented on torch_xla/experimental/custom_kernel.py:
```diff
@@ -331,6 +331,24 @@ def flash_attention(
   return FlashAttention.apply(q, k, v, causal)
 
 
+def paged_attention(q, k_pages, v_pages, lengths, page_indices,
```
how do you test this method? I see you are testing the functionality directly - via bypassing this method.

wonjoolee95 replied:
Yep, we'll have two separate unit tests: one with directly bypassing this method and another one to test this wrapper. I updated PR with the latter.

miladm commented:
cc @WoosukKwon

wonjoolee95 commented:
Locally, the tests are succeeding on my v4:
```
root@1fdc3324aeef:/pytorch/xla# python test/test_pallas.py PallasTest.test_paged_attention_wrapper
.
----------------------------------------------------------------------
Ran 1 test in 2.209s

OK
root@1fdc3324aeef:/pytorch/xla# python test/test_pallas.py PallasTest.test_paged_attention_wrapper_with_dynamo
.
----------------------------------------------------------------------
Ran 1 test in 2.114s

OK
```

I also just triggered the TPU CI on this PR.

wonjoolee95 commented:
The CPU CI is failing with an unrelated test:
```
======================================================================
FAIL: test_resnet18 (__main__.DynamoTrainingBasicTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/xla/xla/pytorch/xla/test/dynamo/test_dynamo.py", line 494, in test_resnet18
    self.assertEqual(met.metric_data('ExecuteTime')[0], sample_count * 3)
AssertionError: 29 != 30
```
The CI including the TPU CI is passing, so this PR should be good to be reviewed. A gentle ping, cc @alanwaketan 😄  Thanks!

miladm reviewed (approved):
Thanks @wonjoolee95  - left a comment for you to eval and address - approving to unblock you

alanwaketan commented on torch_xla/experimental/custom_kernel.py:
```diff
@@ -331,6 +331,51 @@ def flash_attention(
   return FlashAttention.apply(q, k, v, causal)
 
 
+def paged_attention(q, k_pages, v_pages, lengths, page_indices,
```
The original kernel has this thing called: q_dtype_for_kernel_launch? What does it do? Should we copy that as well?

wonjoolee95 replied:
In the original kernel, the `q_dtype_for_kernel_launch` is always either `jnp.float32` or q's `dtype`. In our case, I'm expecting the passed-in q's `dtype` to be `torch.32`, so the `q_dtype_for_kernel_launch` will always be `float32`.

alanwaketan replied:
No, I don't think that will be the case for actual workflow. It could be bf16 or even in8 etc...

wonjoolee95 replied:
I see, makes sense. Just updated to handle `q_dtype_for_kernel_launch`, following jax's kernel -- https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/tpu/paged_attention/paged_attention_kernel.py#L393. I can follow-up in another PR to add some more unit tests for different dtypes for `q`.

alanwaketan commented on torch_xla/experimental/custom_kernel.py:
```diff
@@ -360,3 +405,35 @@ def flash_attention_non_xla(q: torch.Tensor,
+                            pages_per_compute_block: int):
+  # This will be called when dynamo use fake tensor to construct the fake output.
+  # We need to make sure output tensor's shape is correct.
+  if k.device != torch.device("meta"):
```
It feels like this part can be consolidated with the flash attention one.

wonjoolee95 replied:
Sg, refactored these into a helper function.

alanwaketan commented on test/test_pallas.py:
```diff
@@ -477,6 +505,147 @@ def test_flash_attention_backward(self):
+  @unittest.skipIf(xr.device_type() != 'TPU' or tpu.version() < 4,
+                   "This test only works on TPUv4+.")
+  def test_paged_attention_wrapper(self):
+    jax.config.update('jax_default_matmul_precision', jax.lax.Precision.HIGHEST)
```
It's interesting that you use jax as the reference. I guess that works too. Wondering if we can just the eager attention helper in the class instead? Or that doesn't work? Anyway, if you are using jax as the reference, you can drop this.

wonjoolee95 replied:
Sg, yeah I saw that we're dependent on JAX Pallas anyways, so I thought it may be easier to just test against the JAX's outputs. 

Ah, makes sense removed the `jax.config` updates.

alanwaketan commented on test/test_pallas.py:
```diff
@@ -477,6 +505,147 @@ def test_flash_attention_backward(self):
+        q_xla,
+        k_pages_xla,
+        v_pages_xla,
+        seq_lens_xla,
```
Can you explain what these seq_lens are? Are these the previous tokens for each batch in k, v?

wonjoolee95 replied:
Yep, that is my understanding -- the `seq_lens` here equals the number of tokens that are processed in the batch. Reference: https://docs.vllm.ai/en/latest/dev/kernel/paged_attention.html#concepts.

alanwaketan reviewed (approved):
In general, it looks good to me. Left a few comments.

wonjoolee95 commented on test/test_pallas.py:
```diff
@@ -477,6 +505,147 @@ def test_flash_attention_backward(self):
+        torch.allclose(
+            output.cpu()[seq_lens > 0],
+            expected_output.cpu()[seq_lens > 0],
+            atol=1e-1,
```
Sg, updated.

JackCaoG commented on torch_xla/experimental/custom_kernel.py:
```diff
@@ -371,6 +371,65 @@ def flash_attention(
+  # We need to make sure output tensor's shape is correct.
+  if k.device != torch.device("meta"):
+    warnings.warn(
+        'XLA flash attention should only be applied to tensors on XLA device')
```
actually it is not even `paged attention`, you can just make this warning message more general.

wonjoolee95 replied:
Good catch, updated to use an `f` string.

alanwaketan commented on torch_xla/experimental/custom_kernel.py:
```diff
@@ -400,6 +400,9 @@ def paged_attention(q, k_pages, v_pages, lengths, page_indices,
   step = torch.zeros((1,), dtype=torch.int32).to("xla")
   output_shape = torch.Size(list(q.shape[:-1]) + [1])
+  q_output_dtype = torch.float32
+  if (num_heads // num_kv_heads) % 8 != 0:
```
I guess you can combine this with the above L396 code.

wonjoolee95 commented:
Thanks all for the reviews. After addressing all the comments, the two unit tests are still passing locally on my V4. I'll let the TPU CI verify one more time before merging.

alanwaketan commented on torch_xla/experimental/custom_kernel.py:
```diff
@@ -371,6 +371,67 @@ def flash_attention(
+      ], payload, [q.shape, output_shape, output_shape],
+      [q_output_dtype, torch.float32, torch.float32])
+
+  return output.reshape(batch_size, num_heads, head_dim)
```
You probably want to use `.to` to cast the output back to the original dtype here.

wonjoolee95 replied:
Updated.

wonjoolee95 commented:
Merging as all CI is green.
